src/object.d

```diff
@@ -159,15 +170,271 @@ auto opEquals(Object lhs, Object rhs)
     return lhs.opEquals(rhs) && rhs.opEquals(lhs);
 }
 
-/************************
-* Returns true if lhs and rhs are equal.
-*/
-auto opEquals(const Object lhs, const Object rhs)
-{
-    // A hack for the moment.
+bool opEquals(T, U)(const T lhs, const U rhs)
+    if (is(T == class) && is(U == class) &&
+        __traits(hasMember, T, "opEquals") &&
+        __traits(hasMember, U, "opEquals") &&
+        !is(typeof(lhs.opEquals(rhs)) == bool) &&
+        !is(typeof(rhs.opEquals(lhs)) == bool))
+{
+    // FIXME. This is a hack.
+    // We shouldn't need to cast away const, and if either lhs' or rhs' opEquals
+    // mutates either object, it's undefined behavior. But before we can remove
+    // this, we need to make it so that TypeInfo and friends have the corect
+    // definitions for opEquals so that they work with the other overload. And
+    // any user code using const objects but which doesn't define opEquals such
+    // that it works with const with the other overload will also break once
+    // this is removed. So, we need to get rid of this, but we need to be
+    // careful about how and when we do it.
```

Comment:
Has the option of adding a `const` overload of `opEquals` been explored?

Reply:
I don't remember. The decision had been to remove `opEquals` and the other three functions entirely. It's just that we haven't been able to do that, because the built-in AAs use them. The work to templatize them has to be finished first. The other issue of course is that while we decided that the way to fix the attribute problems with these functions was to remove them from `Object`, almost no work has actually gone towards that. AFAIK, this and the work that's been towards fixing the AAs is it. It was a decision made without the work to back it up. And while I originally did the work for this PR years ago, it was blocked due to a compiler bug, and once that was fixed, and I recreated the PR, it just sat here. Adding a `const` overload wouldn't solve the overall attribute problem to allow classes to define these methods with whatever attributes were appropriate, and we were trying to solve that, not `const` specifically. As such, I'm not sure that overloading on `const` was ever really considered, though since one of the goals was to allow for classes to be able to have a non-`const` `opEquals` so that they could do stuff like lazy initialization, adding a `const` overload didn't really fix the problem even when just looking at `const`.

If we ignore the overall attribute problem and just look at `const`, then adding an overload for `opEquals` which is `const` sort of fixes the problem but sort of doesn't (even ignoring the issue of allowing classes to lazy initialize members). If a type just overrides the `const` overload, and `==` or `!=` is used on a mutable object, then the mutable overload in `Object` would be called and not the `const` override. The same goes if a mutable override was declared, and a `const` object were used (in that case, you'd just be using the wrong overload instead of breaking the type system like we do now). Making the mutable `Object.opEquals` call the `const` one would fix the case where a derived class just declared the `const` override, but it wouldn't fix the case where the derived class just declared the mutable override. If we combined that with this template solution, then as long as folks weren't comparing `Object`, that wouldn't be a problem, because the derived class versions would be called directly, but any time that someone (like the current AA implementation) compared `Object`s, then there would be bugs. The result of that approach might be better than what we have now, because it wouldn't break the type system, but it would be buggy for any class that didn't override both overloads of `opEquals`, and it leaves the problem with allowing stuff like lazy initialiazation, but to fix that, `opEquals` has to be removed from `Object`, or we need a different root object for such a class. So, if we add `ProtoObject`, adding a `const` overload to `Object` then just solve the undefined behavior problem, but it causes other bugs unless the compiler forces you to override both overloads, which would be annoying but at least wouldn't be buggy.

The other question is how to then go about deprecating the current behavior. We'd basically have to add some sort of deprecation warning to the compiler about it, and then later change druntime, and break any code that hadn't been updated to have a `const` overload of `opEquals`. But if we were adding that plumbing to the compiler to give a deprecation warning, we could just make it permanently give an error about overriding only one of the two overloads. That's not exactly pretty given that in principle, it should be possible to only declare the `const` version if you don't need them to be different, but it would solve the problem with undefined behavior while making it possible to compare `const Object`s.

Reply:
The most effective is to go with a DIP on `ProtoObject` and then make only non-breaking improvements to `Object`. `ProtoObject` essentially makes `Object` legacy code so its issues become moot.